### ISSUE TITLE
Convergence to a single branch

### ISSUE THREAD

**Issue by kpet** [KhronosGroup/OpenCL-CTS#259]
The medium/long-term plan for the conformance tests is to converge to a single branch that can be used for all versions of OpenCL. This issue will be updated with the latest plan and progress.

- [x] Replace cl22_trunk with a master branch
- [x] Update Travis config
- [x] Reduce diff between cl21_trunk and master to the minimum
- [x] Make it possible to plug offline vendor compilers for the SPIR-V flow on the master branch
- [x] Introduce a framework change to allow specifying the minimal required version for each test
- [x] Merge cl21_trunk into master
- [x] Reduce diff between cl20_trunk and master to the minimum
- [x] Merge cl20_trunk into master
- [x] Figure out what to do next for cl12_trunk
- [x] Get rid of compatibility suites (#494)
- [x] Change common code on master to support both OpenCL 2.x and 1.x
- [ ] Resolve #517 
- [ ] Mark all tests that require 2.x features
- [x] Merge cl12_trunk into master

**Comment by pierremoreau** [comment]
I started some work on cleaning up the CMake files and allowing the user to select, from CMake, which OpenCL API version to test (as we will want to change the definition of `CL_TARGET_OPENCL_VERSION`, and avoid requiring the offline C++ compiler and libclcxx for versions <= 2.1). I’ll post a MR once I am further along.

**Comment by kpet** [comment]
1. Maybe. I'm hopeful we'll be able to avoid building for a specific version but I'm not 100% sure.
2. Yes, this will need to happen but only right before the merge for each branch so we keep the diff minimal.

**Comment by kpet** [comment]
`cl21_trunk` now merged into `master` \o/.

**Comment by kpet** [comment]
`cl20_trunk` now merged into `master`.

**Comment by kpet** [comment]
Regarding the merge of `cl12_trunk` into `master`, I think the first step is to make it possible to use `master` with a 1.2 implementation. @gwawiork, @alycm is that something you could help with?

**Comment by alycm** [comment]
It's certainly something we'd be interested in having working, I don't have a good handle on how much work it would be though.

To make `master` 1.2 compatible would roughly involve (1) running a 1.2 implementation and seeing what fails, (2) triaging those fails into "tests 2.x feature" and "should not fail", (3) set minimum versions appropriately for the 2.x features, (4) fix the remaining fails.

Step (4) may lead naturally into forward porting any changes that only exist on the `cl12_trunk` branch.

There may well be interference where tests for some 1.x feature now happen to also rely some 2.x feature.  `clCreateCommandQueueWithProperties` is an obvious example, but also an easy one to address.

**Comment by gwawiork** [comment]
I can help. 
I noticed that most kernels are compiled with OCL2.0 build option (to test OpenclC 2.0) and it should be modified in case running on OCL1.2 device (OpenclC 1.2). Cases with read_write image qualifier should be disabled for OCL1.2

**Comment by kpet** [comment]
You're reading my mind :).

> I don't have a good handle on how much work it would be though.

(1), (2) and (3) probably don't require a lot of work. (4) may require substantially more effort but I think we won't get a good idea of the size of it until we start.

` clCreateCommandQueueWithProperties` will definitely be something we have to think about but I've already spotted a couple of other things:
- The compatibility tests. Most of the code is just copy/paste with very few changes and should be removed.
- Build options.  Some tests that existed in the 1.2 branch already now pass `-cl-std=CL2.0` unconditionally.

**Comment by nikhiljnv** [comment]
I think we will need to establish parity between cl12_trunk and master with 1.2 option at some point.
It would make sense to run the tests with both master and cl12_trunk on 1.2  implementation after step 1 and before step 2 as it will help categorize issues against 1.2 requirements (rather than just assessing failures).

**Comment by nikhiljnv** [comment]
> 
> 
> @nikhiljnv I think that is essentially what we're already doing, though that's not what I wrote. We have code that passes cl12_trunk, that we also used to do some quick tests on master.
> 
> I had some slides on this topic at the F2F that we didn't have time to present, but the slides are still in the meeting folder.

Thanks !! Reading through your slides addressed part of my concern. I haven't really looked at the current state of master branch. Will take a look and provide feedback.

**Comment by kpet** [comment]
cl12_trunk is now merged into master. We still need to resolve #517 and mark some of the 2.x suites as requiring 2.x but that shouldn't affect 1.x implementations.
